table/tables/tables.go

```diff
@@ -113,9 +121,47 @@ func TableFromMeta(alloc autoid.Allocator, tblInfo *model.TableInfo) (table.Tabl
 	}
 
 	t.meta = tblInfo
+	partitionExpr, err := generatePartitionExpr(tblInfo)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	t.partitionExpr = partitionExpr
 	return t, nil
 }
 
+func generatePartitionExpr(tblInfo *model.TableInfo) ([]expression.Expression, error) {
+	pi := tblInfo.GetPartitionInfo()
+	if pi == nil {
+		return nil, nil
+	}
+	// TODO: Support other partition method.
+	if pi.Type != model.PartitionTypeRange {
+		// To be compatible with the old code, don't return error here.
+		return nil, nil
+	}
+
+	ctx := mock.NewContext()
```

Comment:
Please do not use `mock.Context`, it should only be used in test.

Reply:
The behavior of this `Context` can be changed to satisfy any test purpose.

Reply:
After a simple discussion, we have come to the agreement that **simple/correctness/performance, here we can't have the cake and eat it too**.
The session context used when creating the table maybe different when user use it,
we must cache the expression for performance, which is the root enemy for correctness.
I'm up for simple but @zz-jason think correctness must comes first.

To archive correctness, what we need is a refactor, pass ctx as parameter: `eval(ctx)`, and expression should not depend on it in the struct define. I wonder would @XuHuaiyu  be glad to do it ?
A compromise is keep it simple is here, we can remove it after the refactor finish.
Anyway, refactor could be parallel and **it should not block table partition implementation**.

Reply:
@tiancaiamao In the discussion, I suggested to **implement another session context** instead of using mock ctx, which **does not** block table partition implementation, and we can removing `ctx` in expression in the future and remove this special, used in table partition only, session context.

Another thing is, if you really thinks **simple** should be the most important thing, I suggest that we no need to buffer the expression, just store the expression string and construct the expression tree every time we executing the query on a partitioned table, this is the **simplest** way and also doesn't have much performance lose.